### PR TITLE
refactor: use promises instead of callbacks

### PR DESCRIPTION
**Pull request by charlierudolph**
@alexdavid @kevgo @alexpeachey

My research into promises

Notes
- really like how it looks. The mixture of async and sync code is far easier and don't have to force sync code into an async interface.
- I don't like bluebird's [promisifyAll](https://github.com/petkaantonov/bluebird/blob/master/API.md#promisepromisifyallobject-target--object-options---object) which you can apply to modules with a lot of functions because it requires adding a suffix. I like the idea of [promisify-node](https://www.npmjs.com/package/promisify-node) and may end up using that. But just went with the simple for now and just promisified the methods I needed
- Unit testing functions that return promises is definitely a bit weird for me. Initially had
  
  ``` coffee
  functionReturningPromise
    .then ((@result) => ), ((@err) => ) # coffeelint required the extra space after the arrow
  ```
  
  Thus did just extended promises in test with
  
  ``` coffee
  functionReturningPromise
    .save @, 'result', 'err'
  ```

### CODE REVIEW

File: features/support/json_helpers.coffee [hunk 1]
```diff
@@ -1,13 +1,16 @@
...
-addToJsonFile = (filePath, toAdd, done) ->
-  fs.readFile filePath, encoding: 'utf8', (err, data) ->
-    output = {}
-    _.assign(output, JSON.parse(data)) unless err
-    _.assign output, toAdd
-    fs.outputJson filePath, output, done
+
+addToJsonFile = (filePath, toAdd) ->
+  readJson filePath
+    .catch -> {}
```
Comment (kevgo) [inline features/support/json_helpers.coffee]: Are you swallowing exceptions here?

File: src/cli.coffee [hunk 2]
```diff
@@ -16,14 +19,13 @@ options = docopt """
...
+  extension = _.find extensions, (ext) -> options[ext]
   src = path.join __dirname, '..', 'config', "default.#{extension}"
   destFilename = "dependency-lint.#{extension}"
   dest = path.join process.cwd(), destFilename
-  callback = (err) ->
-    asyncHandlers.exitOnError err
-    console.log "Configuration file generated at \"#{destFilename}\""
-  fs.copy src, dest, callback
+  copy src, dest
+    .then "Configuration file generated at \"#{destFilename}\""
```
Comment (kevgo) [inline src/cli.coffee]: Doesn't `then` take a function? This is just a string here.

**Comment by kevgo** [comment]
Definitely an interesting change! I'm ambivalent, and happy one way or the other.

I'd say the majority of the simplifications in this PR come from removing the `async` boilerplate (which I never was a fan of). Besides that I see mostly getting rid of `done`, at the cost of higher setup cost (promisifying a lot of things). 

How do you feel about the concerns that (a) not having type information makes it hard to distinguish promises from normal function results, and (b) there is no promise standard in Node, thereby making your API more proprietary?

File: features/support/json_helpers.coffee (hunk 1 above)
Comment (charlierudolph) [inline features/support/json_helpers.coffee]: Yes I am. Was doing that in the previous code. I'll make this better

File: src/cli.coffee (hunk 2 above)
Comment (charlierudolph) [inline src/cli.coffee]: Good catch. fixing

**Comment by charlierudolph** [comment]
To me the benefit comes from the easy in which promise can deal with a mixture of both async and sync code. I used the `async` library to avoid the need for error guards. This always does come at the cost of having to pass around more callbacks.

An example of what I really like is [here](https://github.com/charlierudolph/dependency-lint/pull/8/files#diff-c04ef7d91f80503c677d1df708f61318R18) where the `async.auto` gets simplified with promises and the fact that synchronous code run in the chain of promises gets the try/catch for free preventing any need for wrapping it.

In response to your questions.
a) I don't see any real difference here with the rest of the language. I think you just need to document what returns a promise. Instead of taking a callback as a parameter, which I guess gives a bit of self documentation, you would just need to be explicit about it.

b) Node actually has a basic Promise built in (v0.12 at least) and the major promise libraries all appear to very similar. Most appear to be at least implementing this [standard](https://promisesaplus.com/) and then adding their own convenience methods. For building node modules that would be required in projects using either callbacks or promises, I think you can expose one and have people convert if needed (the big promise libraries provide these methods), or even support both [out of the box](http://mono.software/posts/Creating-NodeJS-modules-with-both-promise-and-callback-API-support-using-Q/). All depends on what you want to write your library

**Comment by kevgo** [comment]
Yeah, that example reads much nicer! I think your promise example is doing things serially now, while the async code did things semi-parallelized. The respective "normal" serialized Node code (with user-friendly error messages, which neither promises nor async allow in this form, but which I think are essential for really good libraries), wouldn't look that much more complex:

``` coffeescript
# E is a helper function that wraps the given high-level error message around the existing error,
# thereby making the error more user friendly as it bubbles up the call chain.
# It then calls 'done'.

find: (dir, done) ->
  E = wrapDone done
  @getModulePackageJsons dir, (e, packageJsons) =>
    if e then return E("Cannot find package JSONs in #{dir}: #{e.message}.\nPlease call DependencyLint in an NPM module's root folder.", e) 
    @getModuleExecutables packageJsons, (e, moduleExecutables) =>
      if e then return E("Error getting module executables in #{e.dirname}: #{e.message}.\nPlease re-run 'npm install' and make sure it finishes without errors.", e) 
      @ensureAllModulesInstalled modulesListed, moduleExecutables, (e) =>
         if e then return E("#{e.module_name} is not properly installed: #{e.message}.\nPlease run 'npm install'.", e) 
         @parseModuleExecutables moduleExecutables, scripts, done
```

How would this look like with promises?

**Comment by kevgo** [comment]
If we don't need meaningful error messages (in prototypes etc), we could extract the repetitive error handling into a helper function. This could look something like this:

``` coffeescript
# errorWrapper returns a function that stands in for async callbacks. 
# If it is given an error, it calls `done` right away.
# Without errors, it calls the given callback with all of the arguments given to itself.
# Below is an untested and probably buggy implementation.
errorWrapper = (done) ->
  (func) ->
    (err) ->
      if err then done err
      func.apply this, arguments

find: (dir, done) ->
  E = errorWrapper done
  @getModulePackageJsons dir, E(_, packageJsons) =>
    @getModuleExecutables packageJsons, E(_, moduleExecutables) =>
      @ensureAllModulesInstalled modulesListed, moduleExecutables, E =>
         @parseModuleExecutables moduleExecutables, scripts, done
```

**Comment by kevgo** [comment]
Mixing sync and async code is easy this way too:

``` coffeescript
E = errorWrapper done
p = async.parallel
syncOp1
asyncOp1 E ->
  syncOp2
  syncOp3
  asyncOp2 E ->
    p [asyncOp3, asyncOp4, asyncOp5], E done
```

**Comment by charlierudolph** [comment]
Yeah that example is almost a waterfall but the need for the `tap` in the middle of the promise chain is why I was using `async.auto`. Refactoring it to use your error wrapper without async it would look like

``` coffee
find: (dir, done) ->
  E = errorWrapper done
  scripts = #...
  modulesListed = #...
  @getModulePackageJsons dir, E (packageJsons) =>
    moduleExecutables = @getModuleExecutables packageJsons
    @ensureAllModulesInstalled {modulesListed, moduleExecutables}, E =>
      done null, @parseModuleExecutables {moduleExecutables, scripts}

# versus promises

find: (dir) ->
  scripts = #...
  modulesListed = #...
  @getModulePackageJsons dir
    .then @getModuleExecutables
    .tap (moduleExecutables) => @ensureAllModulesInstalled {modulesListed, moduleExecutables}
    .then (moduleExecutables) => @parseModuleExecutables {moduleExecutables, scripts}
```

Promises lets you take advantage of return values so never need to do `done null, result`. Then are essentially trading the error wrapper for having to write then at the beginning of each operation. 

In response to the user friendly error messages I think each of those methods should be responsible for making their error messages user friendly instead of having find be responsible for all of them.

The main reason I like using async and now promises is you don't have to indent everything each time you do an async operation. 

I also like how promises can leave error handling as the responsibility of the caller.

**Comment by kevgo** [comment]
Cool stuff here!

Re error messages: if the individual methods are specific for the respective app (and only used once in `find`), then yes, they should provide the error messages. If `find` would call generic methods (fs.read etc), then it needs to create those error messages itself. 

An advantage of explicitly calling `done null, result` is that its clearly visible when a method returns asynchronously, and what exactly it returns. 

I hear you about indentation of async code. I consider it an advantage rather than a disadvantage, because it visualizes the flow of time. I.e. make it clear when things happen right after the previous statement, and when not. Promises hide that somehow. Clearly a matter of taste though. 

I think with the `errorWrapper` helper error handling bubbles up to the caller in both versions, with and without promises.

**Comment by charlierudolph** [comment]
Unit testing becomes nice and easy with [chai-as-promised](https://github.com/domenic/chai-as-promised/). Closing this PR. Uncertain if I'm going to keep promises but leaning towards it at the moment

**Pull request closed without merging**